gogensig:invalid receiver type ..Filename

`llcppg.cfg`
```
{
	"name": "sqlite3",
	"cflags": "$(pkg-config --cflags sqlite3)",
	"libs": "$(pkg-config --libs sqlite3)",
	"include": ["sqlite3.h"],
	"trimPrefixes": ["sqlite3_"],
	"cplusplus": false
}
```
` llcppg -v` got
```bash
2024/12/13 17:48:20 NewFuncDecl: &{sqlite3_open16}
2024/12/13 17:48:20 NewFuncDecl: &{sqlite3_open_v2}
2024/12/13 17:48:20 NewFuncDecl: &{sqlite3_uri_parameter}
panic: -: invalid receiver type ..Filename (..Filename is a pointer type)

goroutine 1 [running]:
github.com/goplus/gogen.(*Package).NewFuncDecl(0x0?, 0x0?, {0x140000abc71, 0xc}, 0x1400000d5c0?)
        /Users/zhangzhiyang/go/pkg/mod/github.com/goplus/gogen@v1.16.3/func.go:139 +0x114
github.com/goplus/llcppg/cmd/gogensig/convert.(*Package).newFuncDeclAndComment(0x140004da570, 0x14000370600, 0x14000353270?, 0x14000353270)
        /Users/zhangzhiyang/Documents/Code/goplus/llcppg/cmd/gogensig/convert/package.go:169 +0x74
github.com/goplus/llcppg/cmd/gogensig/convert.(*Package).NewFuncDecl(0x140004da570, 0x14000353270)
        /Users/zhangzhiyang/Documents/Code/goplus/llcppg/cmd/gogensig/convert/package.go:211 +0x224
github.com/goplus/llcppg/cmd/gogensig/convert.(*AstConvert).VisitFuncDecl(0x101099520?, 0x14000353270)
        /Users/zhangzhiyang/Documents/Code/goplus/llcppg/cmd/gogensig/convert/convert.go:81 +0x24
github.com/goplus/llcppg/cmd/gogensig/visitor.(*BaseDocVisitor).visitFuncDecl(...)
        /Users/zhangzhiyang/Documents/Code/goplus/llcppg/cmd/gogensig/visitor/ast_visitor.go:60
github.com/goplus/llcppg/cmd/gogensig/visitor.(*BaseDocVisitor).visitNode(0x140000b25f0?, {0x1010a39a0?, 0x14000353270?})
        /Users/zhangzhiyang/Documents/Code/goplus/llcppg/cmd/gogensig/visitor/ast_visitor.go:33 +0x94
github.com/goplus/llcppg/cmd/gogensig/visitor.(*BaseDocVisitor).Visit(0x14000133b88?, {0x10108e680?, 0x1400030a280?})
        /Users/zhangzhiyang/Documents/Code/goplus/llcppg/cmd/gogensig/visitor/ast_visitor.go:49 +0x78
github.com/goplus/llcppg/cmd/gogensig/processor.(*DocVisitorManager).Visit(0x14000499c50?, {0x10108e680, 0x1400030a280}, {0x14000026d80, 0x2a}, {0x1400048adf0, 0x9}, 0x60?)
        /Users/zhangzhiyang/Documents/Code/goplus/llcppg/cmd/gogensig/processor/processor.go:23 +0x94
github.com/goplus/llcppg/cmd/gogensig/convert/basic.ConvertProcesser.func1(0x1010a9940?)
        /Users/zhangzhiyang/Documents/Code/goplus/llcppg/cmd/gogensig/convert/basic/basic.go:38 +0x44
github.com/goplus/llcppg/cmd/gogensig/processor.(*DocFileSetProcessor).visitFile(0x1400020e240, {0x14000026d80, 0x2a}, {0x14000499c20?, 0x1, 0x1})
        /Users/zhangzhiyang/Documents/Code/goplus/llcppg/cmd/gogensig/processor/processor.go:77 +0x250
github.com/goplus/llcppg/cmd/gogensig/processor.(*DocFileSetProcessor).ProcessFileSet(0x1400020e240, {0x14000499c20?, 0x1, 0x1})
        /Users/zhangzhiyang/Documents/Code/goplus/llcppg/cmd/gogensig/processor/processor.go:95 +0x144
main.main()
        /Users/zhangzhiyang/Documents/Code/goplus/llcppg/cmd/gogensig/gogensig.go:94 +0x3bc
panic: exit status 2

goroutine 1 [running]:
main.check(...)
        /Users/zhangzhiyang/Documents/Code/goplus/llcppg/llcppg.go:99
main.main()
        /Users/zhangzhiyang/Documents/Code/goplus/llcppg/llcppg.go:94 +0x428
```

--- a/_xtool/llcppsymg/_cmptest/parse_test/llgo.expect
+++ b/_xtool/llcppsymg/_cmptest/parse_test/llgo.expect
@@ -26,6 +26,11 @@ Symbol Map GoName: (*State).Compare, ProtoName In HeaderFile: lua_compare(lua_St
 Symbol Map GoName: (*State).Rawequal, ProtoName In HeaderFile: lua_rawequal(lua_State *, int, int), MangledName: lua_rawequal
 Symbol Map GoName: Sizecomp, ProtoName In HeaderFile: lua_sizecomp(int, int, int, int), MangledName: lua_sizecomp
 
+=== Test Case: InvalidReceiver ===
+Parsed Symbols:
+Symbol Map GoName: (*Sqlite3).Errcode, ProtoName In HeaderFile: sqlite3_errcode(sqlite3 *), MangledName: sqlite3_errcode
+Symbol Map GoName: UriParameter, ProtoName In HeaderFile: sqlite3_uri_parameter(sqlite3_filename, const char *), MangledName: sqlite3_uri_parameter
+
 
 #stderr
 

--- a/_xtool/llcppsymg/_cmptest/parse_test/llgo.expect
+++ b/_xtool/llcppsymg/_cmptest/parse_test/llgo.expect
@@ -35,6 +35,14 @@ Symbol Map GoName: UriParameter, ProtoName In HeaderFile: sqlite3_uri_parameter(
 Parsed Symbols:
 Symbol Map GoName: DerDecoding, ProtoName In HeaderFile: asn1_der_decoding(asn1_node *, const void *, int, char *), MangledName: asn1_der_decoding
 
+=== Test Case: InvalidReceiver typ.NamedType.String is empty ===
+Parsed Symbols:
+Symbol Map GoName: InitWindow, ProtoName In HeaderFile: InitWindow(int, int, const char *), MangledName: InitWindow
+
+=== Test Case: InvalidReceiver typ.canonicalType.Kind == clang.TypePointer ===
+Parsed Symbols:
+Symbol Map GoName: X__MpzSetUiSafe, ProtoName In HeaderFile: __mpz_set_ui_safe(mpz_ptr, unsigned long), MangledName: __mpz_set_ui_safe
+
 
 #stderr
 

--- a/_xtool/llcppsymg/_cmptest/parse_test/llgo.expect
+++ b/_xtool/llcppsymg/_cmptest/parse_test/llgo.expect
@@ -31,6 +31,10 @@ Parsed Symbols:
 Symbol Map GoName: (*Sqlite3).Errcode, ProtoName In HeaderFile: sqlite3_errcode(sqlite3 *), MangledName: sqlite3_errcode
 Symbol Map GoName: UriParameter, ProtoName In HeaderFile: sqlite3_uri_parameter(sqlite3_filename, const char *), MangledName: sqlite3_uri_parameter
 
+=== Test Case: InvalidReceiver PointerLevel > 1 ===
+Parsed Symbols:
+Symbol Map GoName: DerDecoding, ProtoName In HeaderFile: asn1_der_decoding(asn1_node *, const void *, int, char *), MangledName: asn1_der_decoding
+
 
 #stderr
 

--- a/_xtool/llcppsymg/_cmptest/parse_test/parse.go
+++ b/_xtool/llcppsymg/_cmptest/parse_test/parse.go
@@ -118,6 +118,40 @@ int(lua_sizecomp)(size_t s, int idx1, int idx2, int op);
 			isCpp:    false,
 			prefixes: []string{"asn1_"},
 		},
+		{
+			name: "InvalidReceiver typ.NamedType.String is empty",
+			content: `
+			RLAPI void InitWindow(int width, int height, const char *title);
+			`,
+			isCpp:    false,
+			prefixes: []string{""},
+		},
+		{
+			name: "InvalidReceiver typ.canonicalType.Kind == clang.TypePointer",
+			content: `
+			typedef struct
+			{
+			int _mp_alloc;		/* Number of *limbs* allocated and pointed
+							to by the _mp_d field.  */
+			int _mp_size;			/* abs(_mp_size) is the number of limbs the
+							last field points to.  If _mp_size is
+							negative this is a negative number.  */
+			} __mpz_struct;
+			typedef __mpz_struct *mpz_ptr;
+			inline void __mpz_set_ui_safe(mpz_ptr p, unsigned long l)
+{
+  p->_mp_size = (l != 0);
+  p->_mp_d[0] = l & GMP_NUMB_MASK;
+#if __GMPZ_ULI_LIMBS > 1
+  l >>= GMP_NUMB_BITS;
+  p->_mp_d[1] = l;
+  p->_mp_size += (l != 0);
+#endif
+}
+			`,
+			isCpp:    false,
+			prefixes: []string{""},
+		},
 	}
 
 	for _, tc := range testCases {

--- a/_xtool/llcppsymg/_cmptest/parse_test/parse.go
+++ b/_xtool/llcppsymg/_cmptest/parse_test/parse.go
@@ -97,6 +97,17 @@ int(lua_sizecomp)(size_t s, int idx1, int idx2, int op);
 			isCpp:    false,
 			prefixes: []string{"lua_"},
 		},
+		{
+			name: "InvalidReceiver",
+			content: `
+typedef struct sqlite3 sqlite3;
+typedef const char *sqlite3_filename;
+SQLITE_API const char *sqlite3_uri_parameter(sqlite3_filename z, const char *zParam);
+SQLITE_API int sqlite3_errcode(sqlite3 *db);
+            `,
+			isCpp:    false,
+			prefixes: []string{"sqlite3_"},
+		},
 	}
 
 	for _, tc := range testCases {

--- a/_xtool/llcppsymg/_cmptest/parse_test/parse.go
+++ b/_xtool/llcppsymg/_cmptest/parse_test/parse.go
@@ -100,13 +100,23 @@ int(lua_sizecomp)(size_t s, int idx1, int idx2, int op);
 		{
 			name: "InvalidReceiver",
 			content: `
-typedef struct sqlite3 sqlite3;
-typedef const char *sqlite3_filename;
-SQLITE_API const char *sqlite3_uri_parameter(sqlite3_filename z, const char *zParam);
-SQLITE_API int sqlite3_errcode(sqlite3 *db);
-            `,
+			typedef struct sqlite3 sqlite3;
+			typedef const char *sqlite3_filename;
+			SQLITE_API const char *sqlite3_uri_parameter(sqlite3_filename z, const char *zParam);
+			SQLITE_API int sqlite3_errcode(sqlite3 *db);
+			            `,
 			isCpp:    false,
 			prefixes: []string{"sqlite3_"},
+		},
+		{
+			name: "InvalidReceiver PointerLevel > 1",
+			content: `
+			typedef struct asn1_node_st asn1_node_st;
+			typedef asn1_node_st *asn1_node;
+			extern ASN1_API int asn1_der_decoding (asn1_node * element, const void *ider, int ider_len, char *errorDescription);
+						`,
+			isCpp:    false,
+			prefixes: []string{"asn1_"},
 		},
 	}
 

--- a/_xtool/llcppsymg/args/args.go
+++ b/_xtool/llcppsymg/args/args.go
@@ -11,10 +11,11 @@ const LLCPPG_SIGFETCH = "llcppg.sigfetch.json"
 const LLCPPG_PUB = "llcppg.pub"
 
 type Args struct {
-	Help     bool
-	Verbose  bool
-	UseStdin bool
-	CfgFile  string
+	Help                 bool
+	Verbose              bool
+	VerboseParseIsMethod bool //-vpim
+	UseStdin             bool
+	CfgFile              string
 }
 
 func ParseArgs(args []string, defaultCfgFile string, swflags map[string]bool) (*Args, []string) {
@@ -29,6 +30,9 @@ func ParseArgs(args []string, defaultCfgFile string, swflags map[string]bool) (*
 				continue
 			case "-v":
 				result.Verbose = true
+				continue
+			case "-vpim":
+				result.VerboseParseIsMethod = true
 				continue
 			case "-":
 				result.UseStdin = true

--- a/_xtool/llcppsymg/dbg/debug.go
+++ b/_xtool/llcppsymg/dbg/debug.go
@@ -1,0 +1,27 @@
+package dbg
+
+type dbgFlags = int
+
+var flags dbgFlags
+
+const (
+	DbgSymbol        dbgFlags = 1 << iota
+	DbgParseIsMethod          //print parse.go isMethod debug log info
+	DbgFlagAll       = DbgSymbol
+)
+
+func SetDebugSymbol() {
+	flags |= DbgSymbol
+}
+
+func GetDebugSymbol() bool {
+	return flags&DbgSymbol != 0
+}
+
+func SetDebugParseIsMethod() {
+	flags |= DbgParseIsMethod
+}
+
+func GetDebugParseIsMethod() bool {
+	return flags&DbgParseIsMethod != 0
+}

--- a/_xtool/llcppsymg/llcppsymg.go
+++ b/_xtool/llcppsymg/llcppsymg.go
@@ -24,6 +24,7 @@ import (
 	"github.com/goplus/llcppg/_xtool/llcppsymg/args"
 	"github.com/goplus/llcppg/_xtool/llcppsymg/config"
 	"github.com/goplus/llcppg/_xtool/llcppsymg/config/cfgparse"
+	"github.com/goplus/llcppg/_xtool/llcppsymg/dbg"
 	"github.com/goplus/llcppg/_xtool/llcppsymg/parse"
 	"github.com/goplus/llcppg/_xtool/llcppsymg/symbol"
 )
@@ -46,8 +47,12 @@ func main() {
 	check(err)
 	defer conf.Delete()
 
+	if ags.VerboseParseIsMethod {
+		dbg.SetDebugParseIsMethod()
+	}
+
 	if ags.Verbose {
-		symbol.SetDebug(symbol.DbgFlagAll)
+		dbg.SetDebugSymbol()
 		if ags.UseStdin {
 			fmt.Println("Config From Stdin")
 		} else {

--- a/_xtool/llcppsymg/parse/parse.go
+++ b/_xtool/llcppsymg/parse/parse.go
@@ -27,6 +27,14 @@ type SymbolProcessor struct {
 	NameCounts  map[string]int
 }
 
+func panicSourceLocation(loc clang.SourceLocation, prefix string) {
+	fileName := clang.GoString(loc.File().FileName())
+	var line int = int(loc.Line())
+	var column int = int(loc.Column())
+	logString := fmt.Sprintf("########%s#######\nfileName:%s, line:%d, column:%d\n", prefix, fileName, line, column)
+	panic(logString)
+}
+
 func NewSymbolProcessor(Files []string, Prefixes []string) *SymbolProcessor {
 	return &SymbolProcessor{
 		Files:      Files,

--- a/_xtool/llcppsymg/parse/parse.go
+++ b/_xtool/llcppsymg/parse/parse.go
@@ -8,12 +8,11 @@ import (
 	"strings"
 
 	"github.com/goplus/llcppg/_xtool/llcppsymg/clangutils"
+	"github.com/goplus/llcppg/_xtool/llcppsymg/dbg"
 	"github.com/goplus/llcppg/_xtool/llcppsymg/names"
 	"github.com/goplus/llgo/c"
 	"github.com/goplus/llgo/c/clang"
 )
-
-var logIsMethod bool = false
 
 type SymbolInfo struct {
 	GoName    string
@@ -101,7 +100,7 @@ func (p *SymbolProcessor) GenMethodName(class, name string, isDestructor bool, i
 }
 
 func (p *SymbolProcessor) printTypeInfo(typ clang.Type, isArg bool, prefix string) {
-	if logIsMethod {
+	if dbg.GetDebugParseIsMethod() {
 		definitionType := clang.GoString(typ.TypeDeclaration().Definition().Type().String())
 		canonicalType := clang.GoString(typ.CanonicalType().String())
 		fmt.Println("**********", prefix, "**********")
@@ -120,7 +119,7 @@ func (p *SymbolProcessor) printTypeInfo(typ clang.Type, isArg bool, prefix strin
 }
 
 func printResult(isInCurPkg, isPointer bool, goName, prefix string) {
-	if logIsMethod {
+	if dbg.GetDebugParseIsMethod() {
 		fmt.Println("===========", prefix, "===========")
 		fmt.Println("isInCurPkg:", isInCurPkg, "isPointer:", isPointer, "goName", goName)
 	}

--- a/_xtool/llcppsymg/parse/parse.go
+++ b/_xtool/llcppsymg/parse/parse.go
@@ -13,6 +13,8 @@ import (
 	"github.com/goplus/llgo/c/clang"
 )
 
+var logIsMethod bool = false
+
 type SymbolInfo struct {
 	GoName    string
 	ProtoName string
@@ -98,19 +100,60 @@ func (p *SymbolProcessor) GenMethodName(class, name string, isDestructor bool, i
 	return prefix + name
 }
 
+func (p *SymbolProcessor) printTypeInfo(typ clang.Type, isArg bool, prefix string) {
+	if logIsMethod {
+		definitionType := clang.GoString(typ.TypeDeclaration().Definition().Type().String())
+		canonicalType := clang.GoString(typ.CanonicalType().String())
+		fmt.Println("**********", prefix, "**********")
+		fmt.Println(
+			"typ.String():", clang.GoString(typ.String()),
+			"typ.NamedType().String():", clang.GoString(typ.NamedType().String()),
+			"isTypePointer:", typ.Kind == clang.TypePointer,
+			"isTypeElaborated:", typ.Kind == clang.TypeElaborated,
+			"isTypeTypedef:", typ.Kind == clang.TypeTypedef,
+			"isInCurPkg:", p.inCurPkg(typ.TypeDeclaration(), isArg),
+			"definitionType:", definitionType,
+			"canonicalType:", canonicalType,
+			"pointLevel:", p.pointerLevel(typ),
+		)
+	}
+}
+
+func printResult(isInCurPkg, isPointer bool, goName, prefix string) {
+	if logIsMethod {
+		fmt.Println("===========", prefix, "===========")
+		fmt.Println("isInCurPkg:", isInCurPkg, "isPointer:", isPointer, "goName", goName)
+	}
+}
+
+func (p *SymbolProcessor) pointerLevel(typ clang.Type) int {
+	canonicalTypeGoString := clang.GoString(typ.CanonicalType().String())
+	return strings.Count(canonicalTypeGoString, "*")
+}
+
 func (p *SymbolProcessor) isMethod(cur clang.Cursor, isArg bool) (bool, bool, string) {
-	isInCurPkg := p.inCurPkg(cur, isArg)
 	typ := cur.Type()
+	if p.pointerLevel(typ) > 1 {
+		return false, false, names.GoName(clang.GoString(typ.String()), p.Prefixes, false)
+	}
+	isInCurPkg := p.inCurPkg(cur, isArg)
+	p.printTypeInfo(typ, isArg, "typ")
 	if typ.Kind == clang.TypePointer {
 		pointeeType := typ.PointeeType()
-		namedTypeGoString := clang.GoString(pointeeType.NamedType().String())
+		p.printTypeInfo(pointeeType, isArg, "typ.PointeeType()")
+		pointeeTypeNamedType := pointeeType.NamedType()
+		namedTypeGoString := clang.GoString(pointeeTypeNamedType.String())
+		p.printTypeInfo(pointeeTypeNamedType, isArg, "typ.PointeeType().NamedType()")
 		if len(namedTypeGoString) > 0 {
-			return isInCurPkg, true, names.GoName(namedTypeGoString, p.Prefixes, isInCurPkg)
+			goName := names.GoName(namedTypeGoString, p.Prefixes, isInCurPkg)
+			printResult(isInCurPkg, true, goName, "typ.pointeeType().NamedType()")
+			return isInCurPkg, true, goName
 		}
 		return p.isMethod(pointeeType.TypeDeclaration(), isArg)
 	} else if typ.Kind == clang.TypeElaborated ||
 		typ.Kind == clang.TypeTypedef {
 		canonicalType := typ.CanonicalType()
+		p.printTypeInfo(canonicalType, isArg, "typ.CanonicalType()")
 		if canonicalType.Kind == clang.TypePointer {
 			return p.isMethod(canonicalType.TypeDeclaration(), isArg)
 		}
@@ -118,10 +161,14 @@ func (p *SymbolProcessor) isMethod(cur clang.Cursor, isArg bool) (bool, bool, st
 	namedType := typ.NamedType()
 	namedTypeGoString := clang.GoString(namedType.String())
 	if len(namedTypeGoString) > 0 {
-		return isInCurPkg, false, names.GoName(namedTypeGoString, p.Prefixes, isInCurPkg)
+		goName := names.GoName(namedTypeGoString, p.Prefixes, isInCurPkg)
+		printResult(isInCurPkg, false, goName, "typ.NamedType()")
+		return isInCurPkg, false, goName
 	}
 	typeGoString := clang.GoString(typ.String())
-	return isInCurPkg, false, names.GoName(typeGoString, p.Prefixes, isInCurPkg)
+	goName := names.GoName(typeGoString, p.Prefixes, isInCurPkg)
+	printResult(isInCurPkg, false, goName, "typ")
+	return isInCurPkg, false, goName
 }
 
 func (p *SymbolProcessor) genGoName(cursor clang.Cursor) string {

--- a/_xtool/llcppsymg/parse/parse.go
+++ b/_xtool/llcppsymg/parse/parse.go
@@ -104,6 +104,12 @@ func (p *SymbolProcessor) isMethod(cur clang.Cursor, isArg bool) (bool, bool, st
 	if typ.Kind == clang.TypePointer {
 		namedType := typ.PointeeType().NamedType().String()
 		return isInCurPkg, true, names.GoName(clang.GoString(namedType), p.Prefixes, isInCurPkg)
+	} else if typ.Kind == clang.TypeElaborated ||
+		typ.Kind == clang.TypeTypedef {
+		canonicalType := typ.CanonicalType()
+		if canonicalType.Kind == clang.TypePointer {
+			return p.isMethod(canonicalType.TypeDeclaration(), isArg)
+		}
 	}
 	namedType := typ.NamedType().String()
 	return isInCurPkg, false, names.GoName(clang.GoString(namedType), p.Prefixes, isInCurPkg)

--- a/_xtool/llcppsymg/parse/parse.go
+++ b/_xtool/llcppsymg/parse/parse.go
@@ -102,8 +102,12 @@ func (p *SymbolProcessor) isMethod(cur clang.Cursor, isArg bool) (bool, bool, st
 	isInCurPkg := p.inCurPkg(cur, isArg)
 	typ := cur.Type()
 	if typ.Kind == clang.TypePointer {
-		namedType := typ.PointeeType().NamedType().String()
-		return isInCurPkg, true, names.GoName(clang.GoString(namedType), p.Prefixes, isInCurPkg)
+		pointeeType := typ.PointeeType()
+		namedTypeGoString := clang.GoString(pointeeType.NamedType().String())
+		if len(namedTypeGoString) > 0 {
+			return isInCurPkg, true, names.GoName(namedTypeGoString, p.Prefixes, isInCurPkg)
+		}
+		return p.isMethod(pointeeType.TypeDeclaration(), isArg)
 	} else if typ.Kind == clang.TypeElaborated ||
 		typ.Kind == clang.TypeTypedef {
 		canonicalType := typ.CanonicalType()
@@ -111,8 +115,13 @@ func (p *SymbolProcessor) isMethod(cur clang.Cursor, isArg bool) (bool, bool, st
 			return p.isMethod(canonicalType.TypeDeclaration(), isArg)
 		}
 	}
-	namedType := typ.NamedType().String()
-	return isInCurPkg, false, names.GoName(clang.GoString(namedType), p.Prefixes, isInCurPkg)
+	namedType := typ.NamedType()
+	namedTypeGoString := clang.GoString(namedType.String())
+	if len(namedTypeGoString) > 0 {
+		return isInCurPkg, false, names.GoName(namedTypeGoString, p.Prefixes, isInCurPkg)
+	}
+	typeGoString := clang.GoString(typ.String())
+	return isInCurPkg, false, names.GoName(typeGoString, p.Prefixes, isInCurPkg)
 }
 
 func (p *SymbolProcessor) genGoName(cursor clang.Cursor) string {

--- a/_xtool/llcppsymg/symbol/symbol.go
+++ b/_xtool/llcppsymg/symbol/symbol.go
@@ -10,27 +10,13 @@ import (
 
 	"github.com/goplus/llcppg/_xtool/llcppsymg/config"
 	"github.com/goplus/llcppg/_xtool/llcppsymg/config/cfgparse"
+	"github.com/goplus/llcppg/_xtool/llcppsymg/dbg"
 	"github.com/goplus/llcppg/_xtool/llcppsymg/parse"
 	"github.com/goplus/llcppg/types"
 	"github.com/goplus/llgo/c"
 	"github.com/goplus/llgo/c/cjson"
 	"github.com/goplus/llgo/xtool/nm"
 )
-
-type dbgFlags = int
-
-const (
-	DbgSymbol  dbgFlags = 1 << iota
-	DbgFlagAll          = DbgSymbol
-)
-
-var (
-	debugSymbol bool
-)
-
-func SetDebug(dbgFlags dbgFlags) {
-	debugSymbol = (dbgFlags & DbgSymbol) != 0
-}
 
 // ParseDylibSymbols parses symbols from dynamic libraries specified in the lib string.
 // It handles multiple libraries (e.g., -L/opt/homebrew/lib -llua -lm) and returns
@@ -39,12 +25,12 @@ func SetDebug(dbgFlags dbgFlags) {
 //
 // Returns symbols and nil error if any symbols are found, or nil and error if none found.
 func ParseDylibSymbols(lib string) ([]*nm.Symbol, error) {
-	if debugSymbol {
+	if dbg.GetDebugSymbol() {
 		fmt.Println("ParseDylibSymbols:from", lib)
 	}
 	sysPaths := getSysLibPaths()
 	lbs := cfgparse.ParseLibs(lib)
-	if debugSymbol {
+	if dbg.GetDebugSymbol() {
 		fmt.Println("ParseDylibSymbols:LibConfig Parse To")
 		fmt.Println("libs.Names: ", lbs.Names)
 		fmt.Println("libs.Paths: ", lbs.Paths)
@@ -54,7 +40,7 @@ func ParseDylibSymbols(lib string) ([]*nm.Symbol, error) {
 		return nil, fmt.Errorf("failed to generate some dylib paths: %v", err)
 	}
 
-	if debugSymbol {
+	if dbg.GetDebugSymbol() {
 		fmt.Println("ParseDylibSymbols:dylibPaths", dylibPaths)
 		if len(notFounds) > 0 {
 			fmt.Println("ParseDylibSymbols:not found libname", notFounds)
@@ -68,7 +54,7 @@ func ParseDylibSymbols(lib string) ([]*nm.Symbol, error) {
 
 	for _, dylibPath := range dylibPaths {
 		if _, err := os.Stat(dylibPath); err != nil {
-			if debugSymbol {
+			if dbg.GetDebugSymbol() {
 				fmt.Printf("ParseDylibSymbols:Failed to access dylib %s: %v\n", dylibPath, err)
 			}
 			continue
@@ -86,7 +72,7 @@ func ParseDylibSymbols(lib string) ([]*nm.Symbol, error) {
 	}
 
 	if len(symbols) > 0 {
-		if debugSymbol {
+		if dbg.GetDebugSymbol() {
 			if len(parseErrors) > 0 {
 				fmt.Printf("ParseDylibSymbols:Some libraries could not be parsed: %v\n", parseErrors)
 			}
@@ -101,7 +87,7 @@ func ParseDylibSymbols(lib string) ([]*nm.Symbol, error) {
 func getSysLibPaths() []string {
 	var paths []string
 	if runtime.GOOS == "linux" {
-		if debugSymbol {
+		if dbg.GetDebugSymbol() {
 			fmt.Println("getSysLibPaths:find sys lib path from linux")
 		}
 		paths = []string{
@@ -109,13 +95,13 @@ func getSysLibPaths() []string {
 			"/usr/local/lib",
 		}
 		paths = append(paths, getPath("/etc/ld.so.conf")...)
-		if debugSymbol && len(paths) == 0 {
+		if dbg.GetDebugSymbol() && len(paths) == 0 {
 			fmt.Println("getSysLibPaths:/etc/ld.so.conf havent find any path")
 		}
 		confd := "/etc/ld.so.conf.d"
 		dir, err := os.Stat(confd)
 		if err != nil || !dir.IsDir() {
-			if debugSymbol {
+			if dbg.GetDebugSymbol() {
 				fmt.Println("getSysLibPaths:/etc/ld.so.conf.d not found or not dir")
 			}
 			return paths
@@ -133,7 +119,7 @@ func getSysLibPaths() []string {
 }
 
 func getPath(file string) []string {
-	if debugSymbol {
+	if dbg.GetDebugSymbol() {
 		fmt.Println("getPath:from", file)
 	}
 	var paths []string
@@ -207,23 +193,23 @@ func ReadExistingSymbolTable(fileName string) (map[string]types.SymbolInfo, bool
 
 func GenSymbolTableData(commonSymbols []*types.SymbolInfo, existingSymbols map[string]types.SymbolInfo) ([]byte, error) {
 	if len(existingSymbols) > 0 {
-		if debugSymbol {
+		if dbg.GetDebugSymbol() {
 			fmt.Println("GenSymbolTableData:generate symbol table with exist symbol table")
 		}
 		for i := range commonSymbols {
 			if existingSymbol, exists := existingSymbols[commonSymbols[i].Mangle]; exists && commonSymbols[i].Go != existingSymbol.Go {
-				if debugSymbol {
+				if dbg.GetDebugSymbol() {
 					fmt.Println("symbol", commonSymbols[i].Mangle, "already exist, use exist symbol", existingSymbol.Go)
 				}
 				commonSymbols[i].Go = existingSymbol.Go
 			} else {
-				if debugSymbol {
+				if dbg.GetDebugSymbol() {
 					fmt.Println("new symbol", commonSymbols[i].Mangle, "-", commonSymbols[i].CPP, "-", commonSymbols[i].Go)
 				}
 			}
 		}
 	} else {
-		if debugSymbol {
+		if dbg.GetDebugSymbol() {
 			fmt.Println("GenSymbolTableData:generate symbol table without symbol table")
 			for _, symbol := range commonSymbols {
 				fmt.Println("new symbol", symbol.Mangle, "-", symbol.CPP, "-", symbol.Go)
@@ -253,12 +239,12 @@ func GenSymbolTableData(commonSymbols []*types.SymbolInfo, existingSymbols map[s
 
 func GenerateAndUpdateSymbolTable(symbols []*nm.Symbol, headerInfos map[string]*parse.SymbolInfo, symbFile string) ([]byte, error) {
 	commonSymbols := GetCommonSymbols(symbols, headerInfos)
-	if debugSymbol {
+	if dbg.GetDebugSymbol() {
 		fmt.Println("GenerateAndUpdateSymbolTable:", len(commonSymbols), "common symbols")
 	}
 
 	existSymbols, exist := ReadExistingSymbolTable(symbFile)
-	if exist && debugSymbol {
+	if exist && dbg.GetDebugSymbol() {
 		fmt.Println("GenerateAndUpdateSymbolTable:current path have exist symbol table", symbFile)
 	}
 

